### PR TITLE
New serverless pattern - RabbitMQ to Lambda (Terraform)

### PR DESCRIPTION
**Pull request by MakendranG**
*Issue #, if available:* https://github.com/aws-samples/serverless-patterns/issues/2260

*Description of changes:*

This pattern deploys a Lambda function, a RabbitMQ queue and the IAM permissions required to run the application. RabbitMQ invokes the Lambda function when new messages are available.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rabbitmq-lambda-tf/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# RabbitMQ to AWS Lambda
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:1]:
```suggestion
# Amazon MQ for RabbitMQ to AWS Lambda
```

File: rabbitmq-lambda-tf/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+This pattern deploys a Lambda function, a RabbitMQ queue and the IAM permissions required to run the application. RabbitMQ invokes the Lambda function when new messages are available.
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:3]:
```suggestion
This pattern deploys a Lambda function, an Amazon MQ for RabbitMQ queue and the IAM permissions required to run the application. RabbitMQ invokes the Lambda function when new messages are available.
```

File: rabbitmq-lambda-tf/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+## Deployment Instructions
+
+1. Set up your username and password for RabbitMQ access. In a new terminal window enter:
+    ```
+    aws secretsmanager create-secret --name MQaccess --secret-string '{"username": "your-username", "password": "your-password"}'
+    ```
+1. The command responds with the ARN of your new secret. 
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:24]:
```suggestion
1. Create a new directory, navigate to that directory in a terminal and clone the repository:
```

File: rabbitmq-lambda-tf/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    ``` 
+    git clone https://github.com/aws-samples/serverless-patterns
+    ```
+1. Change directory to this pattern's directory
+    ```
+    cd serverless-patterns/rabbitmq-lambda-tf
+    ```
+1. From the command line, initialize terraform to  to downloads and installs the providers defined in the configuration:
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:32]:
```suggestion
1. From the command line, initialize Terraform to downloads and install the providers defined in the configuration:
```

File: rabbitmq-lambda-tf/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+1. Delete all created resources by terraform
+    ```bash
+    terraform destroy
+    ```
+1. Confirm all created resources has been deleted
+    ```bash
+    terraform show
+    ```
+----
+Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:77]:
```suggestion
Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: rabbitmq-lambda-tf/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+1. In the MQconsumer Lambda function, select the Monitoring tab and then choose View logs in CloudWatch. The log streams show that the Lambda function is invoked by Amazon MQ and you see the message in the logs.
+
+## Cleanup
+
+1. Change directory to the pattern directory:
+    ```
+    cd rabbitmq-lambda-tf
+    ```
+1. Delete all created resources by terraform
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/README.md:68]:
```suggestion
1. Delete all created resources by Terraform
```

File: rabbitmq-lambda-tf/example-pattern.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
+{
+  "title": "RabbitMQ to AWS Lambda",
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/example-pattern.json:2]:
```suggestion
  "title": "Amazon MQ for RabbitMQ to AWS Lambda",
```

File: rabbitmq-lambda-tf/example-pattern.json [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		        ]
+	},
+	"testing": {
+		"text": [
+		"See the Github repo for detailed testing instructions."
+		]
+	},
+	"cleanup": {
+		"text": [
+		          "terraform destroy",
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/example-pattern.json:47]:
```suggestion
		          "<code>terraform destroy</code>",
```

File: rabbitmq-lambda-tf/example-pattern.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		          "terraform show"
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/example-pattern.json:48]: Is that really necessary?

File: rabbitmq-lambda-tf/main.tf [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+          "ec2:DescribeNetworkInterfaces",
+          "ec2:DescribeVpcs",
+          "ec2:DeleteNetworkInterface",
+          "ec2:DescribeSubnets",
+          "ec2:DescribeSecurityGroups",
+          "logs:CreateLogGroup",
+          "logs:CreateLogStream",
+          "logs:PutLogEvents"
+        ],
+        Resource = "*"
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/main.tf:87]: Try to apply least privilege. At least restrict them to the current region and account

File: rabbitmq-lambda-tf/main.tf [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  function_name    = aws_lambda_function.mq_consumer.arn
+  source_access_configuration {
+    type = "BASIC_AUTH"
+    uri  = data.aws_secretsmanager_secret.mq_access.arn
+  }
+  queues = ["myQueue"]
+}
+
+
+
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/main.tf:107]: Delete empty lines

File: rabbitmq-lambda-tf/main.tf [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  deployment_mode       = "SINGLE_INSTANCE"
+  
+  user {
+    username = jsondecode(data.aws_secretsmanager_secret_version.mq_access_version.secret_string)["username"]
+    password = jsondecode(data.aws_secretsmanager_secret_version.mq_access_version.secret_string)["password"]
+  }
+}
+
+resource "aws_lambda_function" "mq_consumer" {
+  filename      = "src/app.zip"
```
Comment (bfreiberg) [inline rabbitmq-lambda-tf/main.tf:29]: Is there another way to include the code that is not a zip file?

File: rabbitmq-lambda-tf/main.tf (hunk 10 above)
Comment (MakendranG) [reply rabbitmq-lambda-tf/main.tf]: Modified the code with the least privilege

File: rabbitmq-lambda-tf/main.tf (hunk 11 above)
Comment (MakendranG) [reply rabbitmq-lambda-tf/main.tf]: Removed empty lines

File: rabbitmq-lambda-tf/example-pattern.json (hunk 9 above)
Comment (MakendranG) [reply rabbitmq-lambda-tf/example-pattern.json:48]: Yes, it is needed to confirm there are no resources exist after issuing the command terraform destroy.

**Review by MakendranG — COMMENTED** [review]
As per suggestion, made all changes.

**Comment by bfreiberg** [comment]
Thanks @MakendranG for your contribution. Your PR will be merged to serverlessland.com soon

**Pull request merged**